larray/core.py

```diff
@@ -4685,6 +4685,78 @@ def rename(self, renames=None, to=None, inplace=False, **kwargs):
         else:
             return LArray(self.data, axes)
 
+    def reindex_axis(self, axis, labels, fill_value=np.nan, inplace=False):
+        """Change order of labels of an axis and/or add new ones.
+
+        Place NaN or given `fill_value` in locations having no value previously.
+
+        Parameters
+        ----------
+        axis : int, str or Axis
+            Axis along with to reindex.
+        labels : list of str
+            New list of labels.
+        fill_value : scalar, optional
+            Value set to data corresponding to added labels.
+            By defaults NaN.
+        inplace : bool, optional
+            Whether or not to modify the original object or return a new array and leave the original intact.
+            By defaults False.
+
+        Returns
+        -------
+        LArray
+            Array with reindexed axis.
+
+        Notes
+        -----
+        When introducing NAs into an array containing integers via reindex_axis,
+        all data will be promoted to float in order to store the NAs.
+
+        Examples
+        --------
+        >>> arr = ndtest((2, 2, 2))
+        >>> arr
+         a | b\\c | c0 | c1
+        a0 |  b0 |  0 |  1
+        a0 |  b1 |  2 |  3
+        a1 |  b0 |  4 |  5
+        a1 |  b1 |  6 |  7
+        >>> arr.reindex_axis(x.b, ['b1', 'b2', 'b0'], fill_value=-1)
+         a | b\\c | c0 | c1
+        a0 |  b1 |  2 |  3
+        a0 |  b2 | -1 | -1
+        a0 |  b0 |  0 |  1
+        a1 |  b1 |  6 |  7
+        a1 |  b2 | -1 | -1
+        a1 |  b0 |  4 |  5
+        >>> arr.reindex_axis(x.b, ['b1', 'b2', 'b0'])
+         a | b\c |  c0 |  c1
+        a0 |  b1 | 2.0 | 3.0
+        a0 |  b2 | nan | nan
+        a0 |  b0 | 0.0 | 1.0
+        a1 |  b1 | 6.0 | 7.0
+        a1 |  b2 | nan | nan
+        a1 |  b0 | 4.0 | 5.0
+        """
+        axis = self.axes[axis]
+        if not np.isscalar(fill_value):
+            raise TypeError("fill_value must be a scalar")
+        if np.isnan(fill_value):
+            object.__setattr__(self, 'data', self.data.astype(float))
```

Comment:
in any case, I *think* it shouldn't be necessary to promote explicitly because .extend should handle mixed dtype fine (and promote to the "larger" type)